cloud/scope/vpc_cluster.go

```diff
@@ -686,3 +742,321 @@ func (s *VPCClusterScope) buildCOSObjectHRef() (*string, error) {
+	if err != nil {
+		return fmt.Errorf("error retrieving resource group id for subnet creation: %w", err)
+	} else if resourceGroupID == "" {
+		return fmt.Errorf("error retrieving resource group id for resource group %s", s.IBMVPCCluster.Spec.ResourceGroup)
+	}
+
+	vpcID, err := s.GetVPCID()
+	if err != nil {
+		return fmt.Errorf("error retrieving vpc id for subnet creation: %w", err)
+	}
+
+	// TODO(cjschaef): Move to webhook validation.
+	if subnet.Zone == nil {
+		return fmt.Errorf("error subnet zone must be defined for subnet %s", *subnet.Name)
+	}
+
+	// NOTE(cjschaef): We likely will want to add support to use custom Address Prefixes
+	// For now, we rely on the API to assign us prefixes, as we request via IP count
+	var ipCount int64 = 256
+	// We currnetly only support IPv4
+	ipVersion := "ipv4"
+
+	// Find or create a Public Gateway in this zone for the subnet, only one Public Gateway is required for each zone, for this cluster.
+	// NOTE(cjschaef): We may need to add support to not attach Public Gateways to subnets.
+	publicGateway, err := s.findOrCreatePublicGateway(*subnet.Zone)
+	if err != nil {
+		return fmt.Errorf("error failed to find or create public gateway for subnet %s: %w", *subnet.Name, err)
+	}
+
+	options := &vpcv1.CreateSubnetOptions{}
+	options.SetSubnetPrototype(&vpcv1.SubnetPrototype{
+		IPVersion:             ptr.To(ipVersion),
+		TotalIpv4AddressCount: ptr.To(ipCount),
+		Name:                  subnet.Name,
+		VPC: &vpcv1.VPCIdentity{
+			ID: vpcID,
+		},
+		Zone: &vpcv1.ZoneIdentity{
+			Name: subnet.Zone,
+		},
+		ResourceGroup: &vpcv1.ResourceGroupIdentity{
+			ID: ptr.To(resourceGroupID),
+		},
+		PublicGateway: &vpcv1.PublicGatewayIdentity{
+			ID: publicGateway.ID,
+		},
+	})
+
+	// Create subnet.
+	subnetDetails, _, err := s.VPCClient.CreateSubnet(options)
+	if err != nil {
+		return fmt.Errorf("error unknown failure creating vpc subnet: %w", err)
+	}
+	if subnetDetails == nil || subnetDetails.ID == nil || subnetDetails.CRN == nil {
+		return fmt.Errorf("error failed creating subnet: %s", *subnet.Name)
+	}
+
+	// Initially populate subnet's status.
+	resourceStatus := &infrav1beta2.ResourceStatus{
+		ID:    *subnetDetails.ID,
+		Name:  subnetDetails.Name,
+		Ready: false,
+	}
+	if isControlPlane {
+		s.SetResourceStatus(infrav1beta2.ResourceTypeControlPlaneSubnet, resourceStatus)
+	} else {
+		s.SetResourceStatus(infrav1beta2.ResourceTypeWorkerSubnet, resourceStatus)
+	}
+
+	// Add a tag to the subnet for the cluster.
+	err = s.TagResource(s.IBMVPCCluster.Name, *subnetDetails.CRN)
+	if err != nil {
+		return fmt.Errorf("error failed to tag subnet %s: %w", *subnetDetails.Name, err)
+	}
+
+	return nil
+}
+
+// findOrCreatePublicGateway will attempt to find if there is an existing Public Gateway for a specific zone, for the cluster (in cluster's Resource Group and VPC), or create a new one. Only one Public Gateway is required in each zone, for any subnets in that zone.
```

Comment:
For my better understanding, Do you plan to support multiple subnets in same zone, Do you have any use case in mind when we will need to do this.

Reply:
Yes, the simple path here, is when using different subnets for Control Plane and Data Plane, distributed across the same zones (one CP subnet per zone and one DP subnet per zone).
Otherwise, if the user desires to include edge subnets, or distribute nodes within smaller subnets IP CIDR's, multiple subnets could be used per zone as well.